design/how-ironic-works.md

```diff
@@ -0,0 +1,356 @@
+
+## How ironic writes an operating system image to baremetal?
+
+Ironic supports two fundamental types of disk images.
+
+Whole-disk and Partition (or filesystem) images.
+
+Due to the nature of partition images, where a bootloader and configuration
+must be put in place and configured, it is simpler to focus on the use of
+whole disk images.
+
+The basic workflow consists of:
+
+1. Booting the deployment ramdisk
+2. The deployment ramdisk checks in with the ironic introspection
+   service which updates configuration information stored in ironic
+   about the baremetal machine.
+3. The deployment ramdisk checks in with the ironic service and leverages
+   the MAC addresses to help identify the physical machine in the hardware
+   inventory.
+4. Ironic initiates deployment by first identifying the root disk upon
+   which the disk image is to be written. By default this will be the
+   smallest storage device available, and can be overridden via explict
+   configuration of a [root_device hint][https://docs.openstack.org/ironic/latest/install/advanced.html#specifying-the-disk-for-deployment-root-device-hints]
+5. The deployment ramdisk downloads the image to be written and streams
+   that to the storage device.
+6. If defined as part of the deployment, ironic will add an additional
+   partition for a configuration drive. Ironic will then write the
+   configuraiton drive to disk
+7. Finally ironic reboots the machine.
+
+There are some additional steps that ironic performs, mainly fixing
+partition table data with GPT based partition tables in order to
+prevent issues after deployment, but these steps are incorporated
+as part of the deployment sequence to help ensure that the machine
+will deploy successfully without issues.
+
+## What connectivity is required?
+
+To boot the discovery and deployment image on the baremetal hardware:
+
+* DHCP
+* TFTP (if iPXE is not natively supported by the network interfaces.)
+* HTTP in order to download kernel/ramdisk images via HTTP over a TCP
+  connection.
+
+The discovery and deployment ramdisk image needs to be able to:
+
+* DHCP
+* Resolve DNS
+* Connect to the ironic inspector API endpoint, which operates
+  on port 5050/TCP by default.
+* Connect to the ironic API endpoint, which operates on port
+  6385/TCP by default.
+* The ramdisk will likely need to be able to reach an
+  HTTP(s) endpoint in order to download the image files for
+  deployment.
+* Be accessible on port 9999/TCP. This is used by ironic to issue
+  commands to the running ramdisk.
+
+Between ironic and ironic-inspector:
+
+* Each service must be able to reach the API endpoint for
+  the the other service.
+
+## How-to
+
+### How to discover hardware?
+
+New hardware can be discovered by booting the deployment and discovery
+ramdisk with the "ipa-inspection-callback-url" kernel parameter. This URL
+is used by the agent on the deployment and discovery ramdisk as the endpoint
+to post hardware inventory information to.
```

Comment:
The URL should point to the location of ironic-inspector service?

Reply:
Indeed. I thought that was implied. I'll revise.